provider/oci/storage_volumes.go

```diff
@@ -4,43 +4,567 @@
 package oci
 
 import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
 	"github.com/juju/errors"
+	"github.com/juju/utils/clock"
 
+	"github.com/juju/juju/environs/tags"
+	"github.com/juju/juju/instance"
+	providerCommon "github.com/juju/juju/provider/oci/common"
 	"github.com/juju/juju/storage"
+
+	ociCore "github.com/oracle/oci-go-sdk/core"
 )
 
-type volumeSource struct{}
+func mibToGib(m uint64) uint64 {
+	return (m + 1023) / 1024
+}
+
+type volumeSource struct {
+	env        *Environ
+	envName    string
+	modelUUID  string
+	storageAPI providerCommon.OCIStorageClient
+	computeAPI providerCommon.OCIComputeClient
+	clock      clock.Clock
+}
 
 var _ storage.VolumeSource = (*volumeSource)(nil)
 
+func (v *volumeSource) getVolumeStatus(resourceID *string) (string, error) {
+	request := ociCore.GetVolumeRequest{
+		VolumeId: resourceID,
+	}
+
+	response, err := v.storageAPI.GetVolume(context.Background(), request)
+	if err != nil {
+		if v.env.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("volume not found: %s", *resourceID)
+		} else {
+			return "", err
+		}
+	}
+	return string(response.Volume.LifecycleState), nil
+}
+
+func (v *volumeSource) createVolume(p storage.VolumeParams, instanceMap map[instance.Id]*ociInstance) (_ *storage.Volume, err error) {
+	var details ociCore.CreateVolumeResponse
+	defer func() {
+		if err != nil && details.Id != nil {
+			req := ociCore.DeleteVolumeRequest{
+				VolumeId: details.Id,
+			}
+			response, nestedErr := v.storageAPI.DeleteVolume(context.Background(), req)
+			if nestedErr != nil && !v.env.isNotFound(response.RawResponse) {
+				logger.Warningf("failed to cleanup volume: %s", *details.Id)
```

Comment:
we should probably also log the nestedErr content, since it is being suppressed.